### ISSUE TITLE
20$: Detect dead code and unused variables

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#716]
Create a special diagnostic for dead code and unused variables.

Example:

```js
function f() {
  let x = 69;  // info: unused variable
  return 42;
  g();         // info: unreachable code
}
```

In editors, these diagnostics should be enabled by default. LSP has `DiagnosticTag.Unnecessary`, and VS Code probably has an API too.

In the CLI, these diagnostics should be opt-in, or not part of the exit-fail-on set.

(This might need to be split into two or three tasks.)

**Comment by strager** [comment]
Another example of redundant code: `continue` at the end of a loop:
```js
for (let b of bananas) {
  console.log(b);
  continue;        // redundant
}
```

**Comment by strager** [comment]
Another example of redundant code: `true&&` and `false||`:

`&&true` is not entirely redundant though:
```
> true && "x"
'x'
> false || "x"
'x'
> "x" && true  // <--
true
> "x" || false
'x'
```
Same with `|| false`:
```
> true && null
null
> false || null
null
> null && true
null
> null || false  // <--
false
```

**Comment by strager** [comment]
Another example: `? true : false`

**Comment by strager** [comment]
Another example: returning `undefined` explicitly at the end of a function:
```js
function f() {
  console.log("hi");
  return;             // redundant
}
```

**Comment by strager** [comment]
Another example: chained TypeScript null assertions:
```js
let x: int? = 42;
console.log(x!!!);
//            ^^ redundant

**Comment by strager** [comment]
Another example: empty `export` statements:
```js
export {};  // redundant
```

**Comment by strager** [comment]
Another example: final assignment to a local variable which isn't later referenced:
```js
let x = 42;
x = 69;      // warning
```

**Comment by strager** [comment]
Another example: `return` with value in a setter:
```js
class C {
  get x() { return this._x; }
  set x(v) {
    this._x = v;
    return v;  // warning
  }
}
```

**Comment by strager** [comment]
More examples in TypeScript:
```typescript
x!!       // second ! is redundant
x!?.prop  // ? is redundant
x!?.()    // ?. is redundant
```

**Comment by strager** [comment]
`C` will never get executed in the following code, so we should mark it as dead code:
```javascript
for (A; B; C) {
  D;
  return E;
}
```
